### PR TITLE
shim-sev: Avoid mapping unencrypted and encrypted pages to the same physical memory

### PR DESCRIPTION
**Pull request by haraldh**
Moved some comments to the end of the line to improve readability.

Extended some comments with further information.

Now, that const fn can be used to initialize the static arrays,
and to have less assembler code, this patch moves the static
page tables to rust code.

`array-const-fn-init` is used to initialize the arrays with a
const fn.

Avoid mapping the unencrypted and encrypted pages to the same
physical memory, which according to 7.10.6 of the
AMD64 Architecture Programmer’s Manual Volume 2 is needed.

For the initial page table setup phase this patch:
- adds a PDT_OFFSET table for fine grained page table entries
  for the first 2GB of the encrypted kernel offset address space,
  so that the first pages can be unmapped for unencrypted use
- sets the C-Bit in the PDT_OFFSET table
- also sets the C-Bit in the PDT_IDENT table for 0x00_0000 - 0x20_0000

and after completely switching to kernel virtual address space:
- unmaps 0x00_0000 - 0x20_0000 in the encrypted
  kernel virtual address space
- maps 0x00_0000 - 0x20_0000 from encrypted to unencrypted
  in the identity virtual address space
- unmaps 0x40_0000 - 0x60_0000
  in the identity virtual address space

### CODE REVIEW

**Review by ambaxter — APPROVED** [review]
lgtm

File: internal/shim-sev/src/pagetables.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+//! * PDPT_OFFSET: an offset page table with offset $SHIM_OFFSET
+
+use crate::addr::SHIM_VIRT_OFFSET;
+use crate::paging;
+use array_const_fn_init::array_const_fn_init;
+use x86_64::instructions::tlb::flush;
+use x86_64::structures::paging::{Mapper, Page, PageTableFlags, Size2MiB};
+use x86_64::VirtAddr;
+
+/// an aligned page table page
```
Comment (connorkuehl) [inline internal/shim-sev/src/pagetables.rs:15]:
```suggestion
/// A page-aligned Page Table.
```

File: internal/shim-sev/src/pagetables.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    let step: u64 = 0x200000;
+    base + (i as u64) * step
+}
+
+#[allow(clippy::integer_arithmetic)]
+const fn gen_2gb_pdpt_entries(i: usize) -> u64 {
+    let base: u64 = PageTableFlags::HUGE_PAGE.bits()
+        | PageTableFlags::WRITABLE.bits()
+        | PageTableFlags::PRESENT.bits();
+    let step: u64 = 0x40000000;
```
Comment (connorkuehl) [inline internal/shim-sev/src/pagetables.rs:33]: Same comment about `bytes!` as above

**Review by connorkuehl — CHANGES_REQUESTED** [review]

File: internal/shim-sev/src/pagetables.rs (hunk 1 above)
Comment (haraldh) [reply internal/shim-sev/src/pagetables.rs]: fixed

File: internal/shim-sev/src/pagetables.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+/// an aligned page table page
+#[repr(C, align(4096))]
+pub struct AlignedPageTable(pub [u64; 512]);
+
+#[allow(clippy::integer_arithmetic)]
+const fn gen_2mb_pdt_entries(i: usize) -> u64 {
+    let base: u64 = PageTableFlags::HUGE_PAGE.bits()
+        | PageTableFlags::WRITABLE.bits()
+        | PageTableFlags::PRESENT.bits();
+    let step: u64 = 0x200000;
```
[the remark this answers is not in the document]
Comment (haraldh) [reply internal/shim-sev/src/pagetables.rs]: yes

File: internal/shim-sev/src/pagetables.rs (hunk 2 above)
Comment (haraldh) [reply internal/shim-sev/src/pagetables.rs]: kind of

**Review by npmccallum — CHANGES_REQUESTED** [review]
This commit does three distinct things, but I don't see an obvious dependency between them such that they need to be in a single commit. Can you split this up?

**Comment by haraldh** [comment]
> This commit does three distinct things, but I don't see an obvious dependency between them such that they need to be in a single commit. Can you split this up?

@npmccallum done

**Review by npmccallum — CHANGES_REQUESTED** [review]
Everything looks good except for the commit messages which need to follow:
https://chris.beams.io/posts/git-commit/

**Comment by haraldh** [comment]
> Everything looks good except for the commit messages which need to follow:
> https://chris.beams.io/posts/git-commit/

Extended the commit messages and reworded the initial comment in this PR.

**Review by npmccallum — APPROVED** [review]

**Review by connorkuehl — APPROVED** [review]

**Pull request merged**